share/vss/vss.go

```diff
@@ -0,0 +1,630 @@
+	sessionID []byte
+	// list of deals this Dealer has generated
+	deals []*Deal
+	*aggregator
+}
+
+// Deal encapsulates the verifiable secret share and is sent by the dealer to a verifier.
+type Deal struct {
+	// SessionID is a unique session identifier for this protocol run
+	SessionID []byte
+	// SecShare is the private share generated by the dealer
+	SecShare *share.PriShare
+	// RndShare is the random share generated by the dealer
+	RndShare *share.PriShare
+	// T is the treshold used for this secret sharing run
+	T uint32
+	// Commitments are the coefficients used to verify the shares against.
+	Commitments []abstract.Point
+	// Signature is over the whole deal, and is made using sign.Schnorr. It is
+	// automatically verified by this library.
+	Signature []byte
+}
+
+// Response is sent by the verifiers to all participants and holds each
+// individual validation or refusal of a Deal.
+type Response struct {
+	// SessionID related to this run of the protocol
+	SessionID []byte
+	// Index of the verifier issuing this Response
+	Index uint32
+	// 0 = NO APPROVAL == Complaint , 1 = APPROVAL
+	Status byte
+	// Signature over the whole packet. It is automatically verified by this
+	// library.
+	Signature []byte
+}
+
+const (
+	// StatusComplaint is a constant value meaning that a verifier issues
+	// a Complaint against its Dealer.
+	StatusComplaint byte = iota
+	// StatusApproval is a constant value meaning that a verifier agrees with
+	// the share it received.
+	StatusApproval
+	// special status when a complaint has been justified
+	statusJustified
+)
+
+// Justification is a message that is broadcasted by the Dealer in response to
+// a Complaint. It contains the original Complaint as well as the shares
+// distributed to the complainer.
+type Justification struct {
+	// SessionID related to the current run of the protocol
+	SessionID []byte
+	// Index of the verifier who issued the Complaint,i.e. index of this Deal
+	Index uint32
+	// Deal in clear text
+	Deal *Deal
+	// Signature over the whole packet. It is automatically verified by this
+	// library.
+	Signature []byte
+}
+
+// NewDealer returns a Dealer capable of leading the secret sharing scheme. It
+// does not have to be trusted by other Verifiers. The security parameter t is
+// the number of shares required to reconstruct the secret. It is HIGHLY
+// RECOMMENDED to use a threshold higher or equal than what the method
+// MinimumT() returns, otherwise it breaks the security assumptions of the whole
+// scheme. It returns an error if the t is inferior or equal to 2 or if it is
+// not inferior to t.
+func NewDealer(suite abstract.Suite, longterm, secret abstract.Scalar, verifiers []abstract.Point, r cipher.Stream, t int) (*Dealer, error) {
+	d := &Dealer{
+		suite:     suite,
+		long:      longterm,
+		secret:    secret,
+		verifiers: verifiers,
+	}
+	if !validT(t, verifiers) {
+		return nil, fmt.Errorf("dealer: t %d invalid", t)
+	}
+
+	d.t = t
```

Comment:
because the t is being verified just above. I just prefer setting fields once they're validated even if the end result is exactly the same...